lua/bufferline/get-icon.lua

```diff
@@ -26,7 +26,20 @@ local function get_icon(buffer_name, filetype)
     extension = vim.fn.matchstr(basename, [[\v\.@<=\w+$]], '', '')
   end
 
-  return web.get_icon(basename, extension, { default = true })
+  local iconChar, iconHl = web.get_icon(basename, extension, { default = true })
+
+  if iconHl and vim.fn.hlexists(iconHl..buffer_status) < 1 then
```

Comment:
This creates a custom highlight group for each icon based on the status of the buffer.

The good thing about this approach is that it is status-agnostic: 

* If any more statuses become recognized, it will automatically generate the groups.
* If the highlight groups have already been defined, it won't define them again.
* It uses the group definitions from both the status and icon at runtime, so both this plugin and other colorschemes have a chance to get their foot in the door about what it should look like.

Reply:
I'm wondering, do you have any ideal why this is required on your end, and it works on my end without it? I'm asking because I like to avoid any overhead, performance is important because otherwise switching buffers could become laggy and unpleasant.

Reply:
It has to do with the colorscheme you're using. I'm not sure which one you have, but it's set up in such a way that everything ended up looking nicely. However, on these ones, it did not:

* `nvim-highlite`
* `space-vim-dark`
* The default colorschemes (`default, `blue`, etc)

![cap](https://user-images.githubusercontent.com/36409591/99562640-5a322100-2996-11eb-82ba-72a4316aa371.png)

The ideal for me is that this plugin will be intelligent enough to detect these problems and overcome them, without the user having to know that they existed. 

___

This implementation won't become laggy specifically for one of the reasons I mentioned:

* If the highlight groups have already been defined, it won't define them again.

To prove this, I performed a highly unscientific test by creating 100 different tabs/buffers with 100 different filetypes, and then timed how quick it would be to open all of them and close Neovim. Here are the results:

![cap](https://user-images.githubusercontent.com/36409591/99563649-784c5100-2997-11eb-96c9-e059dd876def.png)

1. The first number is on this PR.
2. The second number is on `master`.

Reply:
:+1: